proposals/306-ipv6-happy-eyeballs.txt

```diff
@@ -48,135 +48,103 @@ Ticket: https://trac.torproject.org/projects/tor/ticket/29801
+3. Relay Connections
 
-   This design attempts a TCP connection to a preferred address. On a failure
-   or a 1.5 second delay, we try the alternative address.
+   If there is an existing authenticated connection, we must use it similar
+   to how we used it pre-Prop306.
 
-   On the first successful TCP connection Tor attempts to authenticate
-   immediately. On the first authentication success, Tor closes the other
-   connection.
+   If there is no existing authenticated connection for an entry node, we
+   must attempt to connect using the first available, allowed, and preferred
+   address.
 
-   This design is the most reliable for clients, but increases the connection
-   load on dual-stack guards and authorities. For instance, this can be used
-   to amplify a DoS attack on the Tor network.
+   We should also allow falling back to the alternate address. For this,
+   a design will be given in Section 3.1.
 
-3.1.3. TCP Connection to Preferred Address On First TCP Success
+3.1. TCP Connection to Preferred Address On First TCP Success
 
-   In this design, we will connect via TCP to the first preferred address. On
-   a failure or after a 1.5 second delay, we attempt to connect via TCP to the
-   alternate address. On a success, Tor attempts to authenticate and closes
-   the other connection.
+   In this design, we will connect via TCP to the first preferred address.
+   On a failure or after a 250 ms delay, we attempt to connect via TCP to
+   the alternate address. On a success, Tor attempts to authenticate and
+   closes the other connection.
 
-   This design is the closest to RFC 8305 and is similar to how Happy Eyeballs
+   This design is close to RFC 8305 and is similar to how Happy Eyeballs
    is implemented in a web browser.
 
-3.2. Recommendations for Implementation of Section 3.1 Proposals
-
-   We should start with implementing and testing the implementation as
-   described in Section 3.1.1 (First Successful Authentication), and then
-   doing the same for the implementations described in 3.1.2 and 3.1.3 if
-   desired or required.
+3.2. Handling Connection Successes And Failures
 
-3.3. Handling Connection Successes And Failures
-
-   Should a connection to a guard succeed and is authenticated via TLS, we
-   can then use the connection. In this case, we should cancel all other
+   Should a connection to a entry node succeed and is authenticated via TLS,
+   we can then use the connection. In this case, we should cancel all other
    connection timers and in-progress connections. Cancelling the timers is
    necessary so we don't attempt new unnecessary connections when our
    existing connection is successful, preventing denial-of-service risks.
 
    However, if we fail all available and allowed connections, we should tell
    the rest of Tor that the connection has failed. This is so we can attempt
-   another guard relay.
+   another entry node.
 
-3.4. Connection Attempt Delays
+3.3. Connection Attempt Delays
```

Comment:
This section is more complicated than it needs to be.

Let's have a single ConnectionAttemptDelay, which defaults to 250ms. Users can adjust it using a torrc option, or we can adjust it network-wide using a consensus parameter.

There should be limits on the ConnectionAttemptDelay:
* the RFC minimum of 10ms (you'll need to add this text)
* a maximum of 30s (as already described in the text)

Reply:
Fixed in 7a947b9.